diff --git a/src/NWints/texas/texas_face.F b/src/NWints/texas/texas_face.F
index e8b52c0a39..8a457080e0 100644
--- a/src/NWints/texas/texas_face.F
+++ b/src/NWints/texas/texas_face.F
@@ -518,7 +518,7 @@ c=================================================================
 c-----------------------------------------------------------------
 c the common block /big/ bl(lcx) has been repleaced by pnl scratch 
 c-----------------------------------------------------------------
-      parameter(mxsh= 190 000)
+      parameter(mxsh= 300 000)
 c------------------------------------
       common /ctxs_index/ maxsh,ifp,inx(mxsh)
       common /inde1/ inxx
diff --git a/src/util/nwc_const.fh b/src/util/nwc_const.fh
index b726362615..939fd5797a 100644
--- a/src/util/nwc_const.fh
+++ b/src/util/nwc_const.fh
@@ -91,7 +91,7 @@
 * Maximum number of shells on a QM system 
 *...................................(must be <= nw_max_atom_shells*nw_max_atom)
       integer nw_max_shells
-      parameter (nw_max_shells = 10000)
+      parameter (nw_max_shells = 15000)
 *------------------------------------------------------------------------------
 * Maximum number of basis functions on an atom 
 *
